src/stackedtime/sparse.jl

```diff
@@ -0,0 +1,199 @@
+##################################################################################
+# This file is part of StateSpaceEcon.jl
+# BSD 3-Clause License
+# Copyright (c) 2020-2023, Bank of Canada
+# All rights reserved.
+##################################################################################
+
+
+# selected sparse linear algebra library is a Symbol
+const sf_libs = (
+    :default,   #
+    :umfpack,   # use Julia's standard library (UMFPACK)
+    :pardiso,   # use Pardiso - the one included with MKL
+)
+
+global sf_default = :umfpack
+
+"""
+    use_umfpack()
+
+Set the default sparse factorization library to UMFPACK (the one used in Julia's
+standard library). See also [`use_pardiso`](@ref).
+
+"""
+@inline use_umfpack() = (global sf_default = :umfpack; nothing)
+# @inline use_umfpack(m::Model) = use_umfpack!(m)
+
+"""
+    use_umfpack!(model)
+
+Instruct the stacked-time solver to use Pardiso with this model. See also
+[`use_pardiso!`](@ref).
+"""
+@inline use_umfpack!(m::Model) = (m.options.factorization = :umfpack; m)
+# @inline use_umfpack!() = use_umfpack()
+export use_umfpack, use_umfpack!
+
+"""
+    use_pardiso()
+
+Set the default sparse factorization library to Pardiso. See also
+[`use_umfpack`](@ref).
+"""
+@inline use_pardiso() = (global sf_default = :pardiso; nothing)
+# @inline use_pardiso(m::Model) = use_pardiso!(m)
+
+"""
+    use_pardiso!(model)
+
+Instruct the stacked-time solver to use Pardiso with this model. 
+"""
+@inline use_pardiso!(m::Model) = (m.options.factorization = :pardiso; m)
+# @inline use_pardiso!() = use_pardiso()
+export use_pardiso, use_pardiso!
+
+### API
+
+# a function to initialize a Factorization instance
+# this is also a good place to do the symbolic analysis
+# sf_prepare(A::SparseMatrixCSC, sparse_lib::Symbol=:default) = sf_prepare(Val(sparse_lib), A)
+# sf_prepare(::Val{S}, args...) where {S} = throw(ArgumentError("Unknown sparse library $S. Try one of $(sf_libs)."))
+
+# a function to calculate the numerical factors
+# sf_factor!(f::Factorization, A::SparseMatrixCSC) = throw(ArgumentError("Unknown factorization type $(typeof(f))."))
+
+# a function to solve the linear system
+# sf_solve!(f::Factorization, x::AbstractArray) = throw(ArgumentError("Unknown factorization type $(typeof(f))."))
+
+
+###########################################################################
+###  Default (UMFPACK)
+sf_prepare(::Val{:default}, A::SparseMatrixCSC) = sf_prepare(Val(sf_default), A)
+
+function _sf_same_sparse_pattern(A::SparseMatrixCSC, B::SparseMatrixCSC)
+    return (A.m == B.m) && (A.n == B.n) && (A.colptr == B.colptr) && (A.rowval == B.rowval)
+end
+
+macro _sf_check_factorize(exception, expression)
+    error = gensym("error")
+    return esc(quote
+        try
+            $expression
+        catch $error
+            if $error isa $exception
```

Comment:
Do we need to pass the exception? Why don't we just have this hard-coded as 
`Union{SingularException,Pardiso.PardisoException,Pardiso.PardisoPosDefException}`

The error message is hard-coded so in theory the exceptions which trigger it should be hard-coded as well.

Reply:
Ok that makes sense. Thank you for explaining.